convert SA controller to shared informers

deads2k opened the pull request:
convert the SA controller to shared informer + workqueue.

I think one of @derekwaynecarr @ncdc or @liggitt

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/35230)

<!-- Reviewable:end -->

deads2k committed:
convert SA controller to shared informers

liggitt commented on cmd/kube-controller-manager/app/controllermanager.go:
```diff
@@ -531,10 +531,11 @@ func StartControllers(s *options.CMServer, kubeconfig *restclient.Config, rootCl
 		client("service-account-controller"),
 		serviceaccountcontroller.DefaultServiceAccountsControllerOptions(),
-	).Run()
+	).Run(5, stop)
```
ConcurrentSASyncs

deads2k replied:
> ConcurrentSASyncs

How about I just set it to one until it gets fixed?

liggitt commented on pkg/controller/serviceaccount/serviceaccounts_controller.go:
```diff
@@ -66,193 +67,159 @@ func DefaultServiceAccountsControllerOptions() ServiceAccountsControllerOptions
+	ns, err := c.nsLister.Get(key)
 	if err != nil {
-		return nil, err
+		return err
```
do we need to tolerate NotFound here? should we add nsSynced to the WaitForCacheSync on startup?

deads2k replied:
> do we need to tolerate NotFound here?

tolerate no. Special handling to skip. sure.



> should we add nsSynced to the WaitForCacheSync on startup?

I don't see a reason to.  It's what we're making primary choices based on.

liggitt commented on pkg/controller/serviceaccount/serviceaccounts_controller.go:
```diff
@@ -66,193 +67,159 @@ func DefaultServiceAccountsControllerOptions() ServiceAccountsControllerOptions
-		return nil, err
-	}
+		if _, err := c.client.Core().ServiceAccounts(ns.Name).Create(&sa); err != nil && !apierrs.IsAlreadyExists(err) {
+			return err
```
short-circuits the whole namespace on one failure?

deads2k replied:
> short-circuits the whole namespace on one failure?

And requeues for retry.  I don't feel strongly though.

liggitt commented on pkg/controller/serviceaccount/serviceaccounts_controller.go:
```diff
@@ -66,193 +67,159 @@ func DefaultServiceAccountsControllerOptions() ServiceAccountsControllerOptions
 	}
-	if len(namespaces) == 1 {
-		return namespaces[0].(*api.Namespace), nil
+	if err != nil {
```
what case is this handling?

deads2k replied:
> what case is this handling?

just cruft.  Will remove.

liggitt commented on pkg/controller/serviceaccount/serviceaccounts_controller.go:
```diff
@@ -66,193 +67,159 @@ func DefaultServiceAccountsControllerOptions() ServiceAccountsControllerOptions
-}
+		// this is only safe because we never read it and we always write it
+		// TODO eliminate this once the fake client can handle creation without NS
+		sa.Namespace = ns.Name
```
yeah, scary looking

deads2k commented:
Addressed comments.  Left one thread like it was before just to avoid having to do some insane plumbing that we don't really care about for this controller.

liggitt commented:
LGTM
